### PR TITLE
initial draft of blueprints

### PR DESCRIPTION
**Pull request by davepacheco**
This PR provides an initial implementation for Blueprints, which we've variously called "update plans" or "deployment plans".  RFD 418 (and the comments here) describe broadly how we expect this to work.

What's here is basically:

- in nexus/types, a few `Blueprint`-related types
- in nexus/deployment, a new module with a low-level builder for making blueprints and the first cut at a higher-level planner
- in the internal API and the Nexus app layer, some basic CRUD-like operations for working with blueprints
- omdb support for the above

What's notably missing:

- public API support (because I think we want to get further along before we expose this to people)
- database support (because that can happen in parallel) -- this stuff is all stored in memory right now
- any automation that uses any of this

Importantly, none of the new stuff in this PR actually does anything unless a person goes out of their way to use the internal APIs to configure it.  Even then, it only generates some structures in memory.  The main reason to land it now is to unblock a bunch of follow-on work like the stuff mentioned above.

A few changes came along for the ride here:

- sled-agent-client and nexus-client use progenitor's "replace" directive for more types that are available in `omicron_common` instead of using the client-generated types.  This is more stuff in the spirit of #4754 but these became surprisingly load-bearing in this case.  A bunch of this PR are changes to a handful of lines in a bunch of files related to this (mostly removing unneeded calls to `into()`, `from()`, `clone()`, or changing an import).
- The simulated sled agent by default now simulates physical disks and zpools.  This support was already there but th default config did not use it.

Some explicit questions for reviewers:

- @ahl I wasn't sure how to structure the various versions of types that we have in nexus/types, especially the API ones
- @leftwo reminded me that we've generally said that omdb will be "do no harm" unless you give it some `-w`/`--destructive` flag (which I don't think exists).  But here I'm adding some commands that do (sort of) modify the system.  I filed #4805 for better omdb support for this sort of thing.  In the meantime, I feel like what's here is a reasonable step (because it doesn't actually do anything yet; when it does start doing something, we may well rip out these interfaces anyway; and I think there's a lot of benefit to having this tooling and I don't know where else to put it).  But if people would rather I left that out I could remove the omdb stuff from this PR.

**Status:** I think the bulk of this is ready for review.  I still want to add a bit to this PR: more automated tests plus some omdb commands for diff'ing a blueprint against another blueprint or an inventory collection.  I hope those things won't change much of what's here but feel free to hold off on review if you're worried about that churn.

### CODE REVIEW

**Comment by davepacheco** [comment]
I incorporated feedback from @ahl and I think I've finished adding all the tests I'm planning to add here.

File: nexus/deployment/src/blueprint_builder.rs [hunk 1]
```diff
@@ -0,0 +1,682 @@
...
+        // The next step is adding these zones to a new sled.
+        let new_sled_id = Uuid::new_v4();
+        let _ = policy_add_sled(&mut policy, new_sled_id);
+        let mut builder =
+            BlueprintBuilder::new_based_on(&blueprint2, &policy, "test_basic");
+        builder.sled_ensure_zone_internal_ntp(new_sled_id).unwrap();
+        let new_sled_resources = policy.sleds.get(&new_sled_id).unwrap();
+        for pool_name in &new_sled_resources.zpools {
+            builder
+                .sled_ensure_zone_crucible(new_sled_id, pool_name.clone())
```
Comment (jgallagher) [inline nexus/deployment/src/blueprint_builder.rs:622]: I'm a tiny bit surprised the builder is okay with "add internal NTP _and_ crucible in one new build", but I guess it's the builder's caller's responsibility to do any kind of dependency sequencing?

File: nexus/deployment/src/planner.rs [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+        // ability to determine whether more internal DNS zones need to be
+        // added and where they should go.  And the blueprint builder will need
+        // to grow the ability to provision one.
+
+        for (sled_id, sled_info) in &self.policy.sleds {
+            // Check for an NTP zone.  Every sled should have one.  If it's not
+            // there, all we can do is provision that one zone.  We have to wait
+            // for that to succeed and synchronize the clock before we can
+            // provision anything else.
+            if self.blueprint.sled_ensure_zone_internal_ntp(*sled_id)? {
```
Comment (jgallagher) [inline nexus/deployment/src/planner.rs:54]: Is this method _slightly_ misnamed because it returns true for boundary NTP too? Maybe just `sled_ensure_zone_ntp` would be more accurate? I assume we will eventually need a check for boundary NTP explicitly (to add more if needed), which might end up making this more confusing.

File: nexus/deployment/src/planner.rs [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+                }
+            }
+
+            if ncrucibles_added > 0 {
+                // Don't make any other changes to this sled.  However, this
+                // change is compatible with any other changes to other sleds,
+                // so we can "continue" here rather than "break".
+                // (Yes, it's currently the last thing in the loop, but being
+                // explicit here means we won't forget to do this when more code
+                // is added below.)
```
Comment (jgallagher) [inline nexus/deployment/src/planner.rs:91]: This implies we should have an explicit `continue` after the `comment` - can we add one?

File: nexus/deployment/src/planner.rs [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+            logctx.log.clone(),
+            &blueprint2,
+            &policy,
+            "test: add NTP?",
+        )
+        .plan()
+        .expect("failed to plan");
+
+        let diff = blueprint2.diff(&blueprint3);
+        println!("2 -> 3 (expect new NTP zone on new sled):\n{}", diff,);
```
Comment (jgallagher) [inline nexus/deployment/src/planner.rs:161]:
```suggestion
        println!("2 -> 3 (expect new NTP zone on new sled):\n{}", diff);
```

File: clients/nexus-client/src/lib.rs [hunk 5]
```diff
@@ -22,6 +22,11 @@ progenitor::generate_api!(
         slog::debug!(log, "client response"; "result" => ?result);
     }),
     replace = {
+        // It's kind of unfortunate to pull in such a complex and unstable type
+        // as "blueprint" this way, but we have really useful functionality
+        // (e.g., diff'ing) that's implemented on our local type.
+        Blueprint = nexus_types::deployment::Blueprint,
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply clients/nexus-client/src/lib.rs:28]: I share yours and Adam's reluctance about `replace`.  In this case, the functionality being shared (diff'ing blueprints) is pretty non-trivial and I think it would be fairly painful to impl the requisite `From`s: I think it would require implementing `From` for `Blueprint`, and `OmicronZonesConfig`, which in turn requires `OmicronZoneConfig`, then `OmicronZoneType` (a big enum with a bunch of properties in many variants), `OmicronZoneDataset`, `SledRole`, `ZpoolName`, `NetworkInterface`, `NetworkInterfaceKind`, and `SourceNatConfig`.  (I'm not positive, but I think that's right.)

I think better approaches here might be to have Progenitor validate that your `replace` type really is compatible with the spec, or alternatively if there were a way to automatically generate `From` impls between two types that are identical (which would presumably break -- correctly -- if they weren't identical).  It doesn't feel sustainable to me to either duplicate complex functionality on multiple types or hand-roll so many complex `From` impls.  What do you think?

File: nexus/deployment/src/planner.rs (hunk 2 above)
Comment (davepacheco) [reply nexus/deployment/src/planner.rs]: Good point.  Changed in c58c7edae to `sled_ensure_zone_ntp`.  (It's a little weird: it's ensuring there's an NTP zone, but always adding an internal one.  I did consider having the planner add boundary ones if needed, but that's more complicated because of #4791.  But I agree this name is less misleading than what was there.)

File: nexus/deployment/src/planner.rs (hunk 3 above)
Comment (davepacheco) [reply nexus/deployment/src/planner.rs:93]: Added in c58c7edae525570e703cb113075211ccd9903c6c.  (I think what happened here is that I talked myself out of needing the `continue` because really you should be able to provision any number of zones at once, once NTP is there (and presuming internal DNS is working).  But rather than overthink it, I'm going to leave the comment and `continue` there and let us revisit when we augment this.)

File: nexus/deployment/src/planner.rs (hunk 4 above)
Comment (davepacheco) [reply nexus/deployment/src/planner.rs]: Fixed in c58c7edae525570e703cb113075211ccd9903c6c.

File: nexus/src/app/deployment.rs [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+        }
+    }
+
+    async fn blueprint_planning_context(
+        &self,
+        opctx: &OpContext,
+    ) -> Result<PlanningContext, Error> {
+        let creator = self.id.to_string();
+        let datastore = self.datastore();
+        let collection = datastore
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/src/app/deployment.rs:205]: Yeah, I avoided thinking too hard about this because it seems like a can of worms and I figured we'd know more about what errors are ambient and problematic vs. not once we had deployed inventory in the wild.  Maybe that point is actually now, but I think what I'd like to do for now is to make the caller (which is basically the person running omdb at this point) specify the collection, leaving it their responsibility to make sure it's okay.  If/when we automate this, we can revisit based on what we found.  Does that sound okay?

File: nexus/src/app/deployment.rs [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+            }
+            zpools
+        };
+
+        let sleds = sled_rows
+            .into_iter()
+            .map(|sled_row| {
+                let sled_id = sled_row.id();
+                let subnet = Ipv6Subnet::<SLED_PREFIX>::new(sled_row.ip());
+                let zpools = zpools_by_sled_id
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/src/app/deployment.rs:262]: I'm on the fence.  Right now that seems likely true.  I'm not sure it always will be.  We may wind up filtering out sleds that are decommissioned.   If we do that, we should probably also exclude their zpools from the query, too.  But that may not be worth the complexity, if decommissioned sleds are pretty uncommon.  Anyway, if somehow this were wrong, I'm not sure we'd want to fail the whole operation.  I could really go either way and I'm usually in favor of extra checks that things aren't off the rails, but my inclination is not to fail in this case because it seems more likely that there's just some runtime condition that we didn't expect that's actually okay than it does that we've got this kind of logic bug.

File: nexus/types/src/deployment.rs [hunk 8]
```diff
@@ -0,0 +1,528 @@
...
+// This is aimed at supporting add/remove sleds.  The plan is to grow this to
+// include more of the system as we support more use cases.
+#[derive(Debug, Clone, Eq, PartialEq, JsonSchema, Deserialize, Serialize)]
+pub struct Blueprint {
+    /// unique identifier for this blueprint
+    pub id: Uuid,
+
+    /// mapping: sled id -> zones deployed on each sled
+    /// A sled is considered part of the control plane cluster iff it has an
+    /// entry in this map.
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/types/src/deployment.rs:111]: I'd say that for me that question is still TBD and it seemed more problematic than useful to add the field at this point, not that it definitely shouldn't be there nor that it definitely will or won't be.

File: nexus/types/src/deployment.rs [hunk 9]
```diff
@@ -0,0 +1,528 @@
...
+                    )
+                })
+                .collect();
+            let mut zones_added = vec![];
+            let mut zones_removed = vec![];
+            let mut zones_changed = vec![];
+
+            // Now go through each zone and compare them.
+            for (zone_id, b1z_info) in &b1zones {
+                if let Some(b2z_info) = b2zones.get(zone_id) {
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/types/src/deployment.rs]: Fixed in c58c7edae525570e703cb113075211ccd9903c6c.  (It was just the one `contains_key`, right?)

File: nexus/types/src/deployment.rs [hunk 10]
```diff
@@ -0,0 +1,528 @@
...
+
+    /// the IPv6 subnet of this sled on the underlay network
+    ///
+    /// (implicitly specifies the whole range of addresses that the planner can
+    /// use for control plane components)
+    pub subnet: Ipv6Subnet<SLED_PREFIX>,
+}
+
+/// Describes a complete set of software and configuration for the system
+// Blueprints are a fundamental part of how the system modifies itself.  Each
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/types/src/deployment.rs:67]: I had that previously.  But this is a publicly-exposed type, so the block comment is used for API consumers, not just Rust consumers.  I remember discussing this months ago and deciding that unless there was a really compelling reason to try to support separate API docs and rustdoc for the same type, we'd just let the rustdoc reflect API docs.

File: nexus/deployment/src/blueprint_builder.rs [hunk 11]
```diff
@@ -0,0 +1,682 @@
...
+            comment: self.comments.join(", "),
+        }
+    }
+
+    /// Sets the blueprints "comment"
+    ///
+    /// This is a short human-readable string summarizing the changes reflected
+    /// in the blueprint.  This is only intended for debugging.
+    pub fn comment(&mut self, comment: &str) {
+        self.comments.push(comment.to_owned());
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/deployment/src/blueprint_builder.rs]: Fixed in c58c7edae525570e703cb113075211ccd9903c6c.

File: nexus/deployment/src/blueprint_builder.rs [hunk 12]
```diff
@@ -0,0 +1,682 @@
...
+    }
+
+    pub fn sled_ensure_zone_internal_ntp(
+        &mut self,
+        sled_id: Uuid,
+    ) -> Result<bool, Error> {
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/deployment/src/blueprint_builder.rs]: Fixed in c58c7edae525570e703cb113075211ccd9903c6c.

File: nexus/deployment/src/blueprint_builder.rs [hunk 13]
```diff
@@ -0,0 +1,682 @@
...
+        // If there's already an NTP zone on this sled, do nothing.
+        let has_ntp = self
+            .parent_blueprint
+            .omicron_zones
+            .get(&sled_id)
+            .map(|found_zones| {
+                found_zones.zones.iter().any(|z| {
+                    matches!(
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/deployment/src/blueprint_builder.rs]: Fixed in c58c7edae525570e703cb113075211ccd9903c6c.

File: nexus/deployment/src/blueprint_builder.rs [hunk 14]
```diff
@@ -0,0 +1,682 @@
...
+        // The list of boundary NTP servers is not necessarily stored
+        // anywhere (unless there happens to be another internal NTP zone
+        // lying around).  Recompute it based on what boundary servers
+        // currently exist.
+        let ntp_servers = self
+            .parent_blueprint
+            .all_omicron_zones()
+            .filter_map(|(_, z)| {
+                if matches!(z.zone_type, OmicronZoneType::BoundaryNtp { .. }) {
+                    Some(format!("{}.host.{}", z.id, DNS_ZONE))
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/deployment/src/blueprint_builder.rs]: Fixed in c58c7edae525570e703cb113075211ccd9903c6c.  There was already some skeleton for this in the `internal-dns`  #crate.  I did have to rework some of this and since they're often used as strings it's not as type-checked as would be nice.  Extra review on this would be helpful (although I expect the end-to-end test job would fail if I somehow got a large class of DNS names wrong).

File: nexus/deployment/src/blueprint_builder.rs [hunk 15]
```diff
@@ -0,0 +1,682 @@
...
+        self.sled_add_zone(sled_id, zone)?;
+        Ok(true)
+    }
+
+    fn sled_add_zone(
+        &mut self,
+        sled_id: Uuid,
+        zone: OmicronZoneConfig,
+    ) -> Result<(), Error> {
+        let _ = self.sled_resources(sled_id)?;
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/deployment/src/blueprint_builder.rs:327]: Added a comment in c58c7edae525570e703cb113075211ccd9903c6c.

File: nexus/deployment/src/blueprint_builder.rs [hunk 16]
```diff
@@ -0,0 +1,682 @@
...
+    fn sled_alloc_ip(&mut self, sled_id: Uuid) -> Result<Ipv6Addr, Error> {
+        let sled_info = self.sled_resources(sled_id)?;
+
+        // Work around rust-lang/rust-clippy#11935 and related issues.
+        // Clippy doesn't grok that we can't use
+        // `entry(&sled_id).or_insert_with(closure)` because `closure` would
+        // borrow `self`, which we can't do because we already have an immutable
+        // borrow of `self` by calling `entry()`.  Using `match` on the result
+        // of `entry` has a similar problem.
+        #[allow(clippy::map_entry)]
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/deployment/src/blueprint_builder.rs]: Thanks!  Done in c58c7edae525570e703cb113075211ccd9903c6c.

File: nexus/deployment/src/blueprint_builder.rs (hunk 1 above)
Comment (davepacheco) [reply nexus/deployment/src/blueprint_builder.rs:630]: Yeah, the basic idea is that the builder is responsible for the mechanism while the planner is responsible for the policy.  It's not so black-and-white, admittedly.  The planner won't let you do things that are easy for it to know cannot work (like putting two zones of the same type on the same zpool).  But "Crucible depends on timesync so it depends on internal NTP and so we don't want to do them in one blueprint" is sufficiently high level that I think the planner should be responsible for it.

File: nexus/deployment/src/blueprint_builder.rs [hunk 17]
```diff
@@ -0,0 +1,690 @@
...
+        // (which itself is a known-prefix parent subnet of the sled subnet).
+        let dns_servers =
+            get_internal_dns_server_addresses(sled_subnet.net().network());
+
+        // The list of boundary NTP servers is not necessarily stored
+        // anywhere (unless there happens to be another internal NTP zone
+        // lying around).  Recompute it based on what boundary servers
+        // currently exist.
+        let ntp_servers = self
+            .parent_blueprint
```
Comment (andrewjstone) [inline nexus/deployment/src/blueprint_builder.rs:246]: This, like other parts of the builder that rely on the `parent_blueprint` seem to implicitly assume that the `parent_blueprint` was successfully executed. That is why they are capable of performing "next steps". Am I correct here?

If I am correct then this may present a problem when a plan is not successfully executed, or partially executed. While conceptually I like the idea of one plan building off the next plan, it seems like we may need to pass in more parameters where this may be problematic, or we need to use the inventory rather than the prior plan.

**Review by andrewjstone — COMMENTED** [review]
Dave, this looks great. It's a huge step forward. I particularly like all the blueprint diff stuff.

File: clients/nexus-client/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (andrewjstone) [reply clients/nexus-client/src/lib.rs:28]: I actually had the opposite reaction, and didn't think it was unfortunate to use `replace` at all. However, clearly I am missing what is problematic here.

[the remark this answers is not in the document]
Comment (davepacheco) [reply clients/nexus-client/src/lib.rs:28]: I think what sucks about using `replace` is that you lose the ability to notice when your local type doesn't match the spec.  In turn that means we lose all of the benefit of Rust's strong static typing across the OpenAPI boundary.

Concretely: in the normal case where Progenitor generates the types for you, say we pull in a new version of the spec with a new required field.  Progenitor's type will have that new field in it.  So if we had a code path that creates an instance of that type (i.e., to make the request) it will now fail to compile.  That's good: if we didn't know it before, we now know we've made an incompatible change to the spec.  If we did know it, now's when we update our code to fill in the new value.

If on the other hand we use `replace` to supply the type, when we pull in the updated spec, everything still compiles -- there's no reason it wouldn't.  Instead, that code path will fail at runtime when the server rejects the request because it's missing the required field.  That really sucks.

Similarly on the output side, if we use `replace` to supply our own type, we don't find out at compile time when our type is incompatible.  A particularly bad case would be an enum, where the server might add a new variant (which one might even consider a non-breaking change).  But if the client sees it, it will fail to deserialize the response.  By contrast, if we use the Progenitor-generated type, we'll fail to build if, say, we try to `match` on the enum and don't include the new variant.

In summary, using Progenitor types gets a lot of the benefit of Rust across the OpenAPI boundary.  Using "replace" bypasses all of it (for that type).  The downside of Progenitor types is that when there's non-trivial functionality on the original type, you either have to impl that functionality twice (sucks) or convert the type to your own type (okay for simple types in a few places, but kind of sucks in the large).  For simple, stable types like `Name` I think the benefits of "replace" probably outweigh the risks.  For complex types that are likely to change, I'm less sure (but still feel what I said earlier about this case).

File: nexus/src/app/deployment.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/src/app/deployment.rs:205]: I want ahead and implemented this in 06825c065.

File: clients/nexus-client/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jgallagher) [reply clients/nexus-client/src/lib.rs:28]: > I think better approaches here might be to have Progenitor validate that your replace type really is compatible with the spec

This feels like it would be huge - it would allow use to use `replace` fearlessly, even on big / complex / volatile types like this one, I think?

File: nexus/types/src/deployment.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (jgallagher) [reply nexus/types/src/deployment.rs:67]: Hmm, but this is an API type for the _internal_ API, right? Do those exist? (To be fair, another question is whether we generate rustdocs for crates. I've done that locally occasionally but it's not a habit.)

File: clients/nexus-client/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (davepacheco) [reply clients/nexus-client/src/lib.rs:28]: > > I think better approaches here might be to have Progenitor validate that your replace type really is compatible with the spec
> 
> This feels like it would be huge - it would allow use to use `replace` fearlessly, even on big / complex / volatile types like this one, I think?

Potentially, yeah. I spoke with Adam about this a bit.  It's not clear how we could do this at compile time, but I imagine we might be able to do it at test-time.

File: nexus/src/app/deployment.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (jgallagher) [reply nexus/src/app/deployment.rs]: Yeah, that seems like a nicer workaround. "I the caller have told you to explicitly use collection X, so presumably I am okay with any errors it reports" feels much more robust than "generate from the latest collection (that might be newer than the last one I saw) and ignore any errors it has".

File: nexus/types/src/deployment.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/types/src/deployment.rs:67]: I've been heavily using API docs for internal APIs (generated by hand with redoc).  Then again I also use rustdocs via rust-analyzer.  But I'm inclined to treat both APIs the same here -- part of the idea behind using OpenAPI everywhere was so that we could have a good experience using our own APIs through the patterns/processes/tooling we use for the external API.  I find it annoying when I'm using the internal API docs and I get a lot of irrelevant Rust-isms.

[the remark this answers is not in the document]
Comment (jgallagher) [reply nexus/types/src/deployment.rs:67]: Fair enough. I don't feel super strongly about this either way, happy to defer to you.

File: nexus/deployment/src/blueprint_builder.rs (hunk 17 above)
Comment (jgallagher) [reply nexus/deployment/src/blueprint_builder.rs:246]: I wondered this too, but wasn't sure how to phrase the question, because I assume it's pretty reasonable we end up in this case:

* We have a current target blueprint that we have not yet achieved
* Something in the system changes such that we can never achieve the current target

Now what? We either need to be able to generate a child blueprint despite the parent not being successfully executed, or maybe we need some escape hatch to say "regenerate a new initial blueprint based on the latest collection"?

File: nexus/src/internal_api/http_entrypoints.rs [hunk 18]
```diff
@@ -591,3 +603,187 @@ async fn ipv4_nat_changeset(
...
+//
+// These are not (yet) intended for use by any other programs.  Eventually, we
+// will want this functionality part of the public API.  But we don't want to
+// commit to any of this yet.  These properly belong in an RFD 399-style
+// "Service and Support API".  Absent that, we stick them here.
+
+/// Lists blueprints
+#[endpoint {
+    method = GET,
+    path = "/deployment/blueprints/all",
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/src/internal_api/http_entrypoints.rs:617]: So right now it's:

```
# list all blueprints
GET /deployment/blueprints/all
# fetch one blueprint
GET /deployment/blueprints/all/$id
# do other stuff, like get the target
GET /deployment/blueprints/target
POST /deployment/blueprints/regenerate
...
```

Basically, if we put the list endpoint at `/deployment/blueprints`, I'd want to put the individual blueprints at `/deployment/blueprints/$id`, but that overlaps with `/deployment/blueprints/target` (and other blueprint-related stuff we might need).  We avoid that (and dropshot doesn't support it, although only because we assumed we'd want to avoid that :)).

I thought we had done this elsewhere in the public API when we need additional control endpoints but I don't actually see any...

Another way to phrase it could be:

```
# list blueprints
GET /deployment/blueprints
# fetch one blueprint
GET /deployment/blueprints/$id
# do other stuff, like get the target
GET /deployment/target-blueprint
POST /deployment/regenerate-blueprint
```

I could go either way.  I'm not too worried at this point as I don't think there will be programmatic consumers of the internal API aside from omdb.

File: nexus/types/src/deployment.rs [hunk 19]
```diff
@@ -0,0 +1,569 @@
...
+use schemars::JsonSchema;
+use serde::Deserialize;
+use serde::Serialize;
+use std::collections::BTreeMap;
+use std::collections::BTreeSet;
+use uuid::Uuid;
+
+/// Fleet-wide deployment policy
+///
+/// The **policy** represents the deployment controls that people (operators and
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/types/src/deployment.rs:32]: Yeah, I anticipate that there will be some user-configurable knobs (like number of Cockroach nodes) and they themselves will have constraints (like "cannot be less than three").

There's also policy that's essentially hardcoded, like "there must be an NTP zone on each sled".

It's not clear to me to what extent the policy would be stored discretely (like, a Policy table or something like that) vs. assembled from existing state.  Probably some of both?  In this PR, the only policy is the set of sleds in service, and that comes from the `sled` table, so it's not like there's a `policy` table even though the policy does come from the database.  But if/when we do support configuring "number of Cockroach nodes", I expect that'll be in some explicit table of user-configured policy knobs.

File: nexus/deployment/src/blueprint_builder.rs (hunk 17 above)
Comment (davepacheco) [reply nexus/deployment/src/blueprint_builder.rs:246]: > I wondered this too, but wasn't sure how to phrase the question, because I assume it's pretty reasonable we end up in this case:
> 
>     * We have a current target blueprint that we have not yet achieved
> 
>     * Something in the system changes such that we can never achieve the current target
> 
> 
> Now what? We either need to be able to generate a child blueprint despite the parent not being successfully executed, or maybe we need some escape hatch to say "regenerate a new initial blueprint based on the latest collection"?

Agreed.  I think the conclusion is that we can't assume the parent blueprint has/will be executed.  So far, I don't think that's been a difficult constraint.

File: nexus/src/internal_api/http_entrypoints.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (andrewjstone) [reply nexus/src/internal_api/http_entrypoints.rs:617]: I think you should keep it as is. Thanks for the explanation!

File: nexus/deployment/src/ip_allocator.rs [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+    pub fn new(minimum: Ipv6Addr, maximum: Ipv6Addr) -> IpAllocator {
+        IpAllocator { last: minimum, maximum }
+    }
+
+    /// Mark the given address reserved so that it will never be returned by
+    /// `alloc()`.
+    ///
+    /// The given address can be outside the range provided to
+    /// `IpAllocator::new()`, in which case this reservation will be ignored.
+    pub fn reserve(&mut self, addr: Ipv6Addr) {
```
[the remark this answers is not in the document]
Comment (andrewjstone) [reply nexus/deployment/src/ip_allocator.rs:35]: That makes sense. Fine to leave it.

File: nexus/deployment/src/blueprint_builder.rs (hunk 17 above)
Comment (andrewjstone) [reply nexus/deployment/src/blueprint_builder.rs:246]: Your third principle covers what I was worried about. Specifically if we started to deploy crucible zones assuming that the internal NTP zone was deployed already. If the planner takes into consideration the current inventory and that the parent blueprint may not have executed before generating the next blueprint, then we should be all set. This also means that the planner may very well be in a "wait for the target blueprint to complete" before it does anything for a good amount of time. That seems normal though.

**Review by andrewjstone — APPROVED** [review]
Ship it!

File: nexus/deployment/src/blueprint_builder.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (jgallagher) [reply nexus/deployment/src/blueprint_builder.rs]: I'm not super familiar with these, but the changes you made look good, and I don't see any lingering, now-suspect `DNS_ZONE` references. :+1:

File: nexus/deployment/src/blueprint_builder.rs [hunk 21]
```diff
@@ -0,0 +1,690 @@
...
+        sled_zones.zones.push(zone);
+        Ok(())
+    }
+
+    /// Returns a newly-allocated underlay address suitable for use by Omicron
+    /// zones
+    fn sled_alloc_ip(&mut self, sled_id: Uuid) -> Result<Ipv6Addr, Error> {
+        let sled_subnet = self.sled_resources(sled_id)?.subnet;
+
+        // Work around rust-lang/rust-clippy#11935 and related issues.
```
Comment (jgallagher) [inline nexus/deployment/src/blueprint_builder.rs:366]: This comment is no longer relevant since we reworked this

**Review by jgallagher — APPROVED** [review]
Changes look great - just one nit about a comment. Thanks!

**Comment by davepacheco** [comment]
Thanks for the close reviews, @jgallagher and @andrewjstone!
